src/test/MockVault.sol

```diff
@@ -4,38 +4,54 @@ pragma solidity ^0.8.0;
 import {PoolId} from "../types/PoolId.sol";
 import {PoolKey} from "../types/PoolKey.sol";
 import {BalanceDelta} from "../types/BalanceDelta.sol";
-import {Currency, CurrencyLibrary} from "../types/Currency.sol";
+import {Currency, CurrencyLibrary, equals as currencyEquals} from "../types/Currency.sol";
 import {SafeCast} from "../libraries/SafeCast.sol";
 
 contract MockVault {
     using SafeCast for *;
     using CurrencyLibrary for Currency;
 
     mapping(address app => mapping(Currency currency => uint256 reserve)) public reservesOfApp;
+
+    // Need to update this when try to record balanceDeltaOfPool
+    PoolKey public currentPoolKey;
     mapping(PoolId poolId => BalanceDelta delta) public balanceDeltaOfPool;
 
+    error InvalidPoolKey();
+
     constructor() {}
 
-    function accountAppBalanceDelta(PoolKey memory key, BalanceDelta delta, address) external {
-        PoolId poolId = key.toId();
-        balanceDeltaOfPool[poolId] = delta;
+    function updateCurrentPoolKey(PoolKey memory key) external {
+        currentPoolKey = key;
+    }
+
+    function accountAppBalanceDelta(Currency currency0, Currency currency1, BalanceDelta delta, address) external {
+        // Will not record balanceDeltaOfPool if currentPoolKey is not set
+        if (!currentPoolKey.currency0.isNative() || !currentPoolKey.currency1.isNative()) {
+            if (
+                !currencyEquals(currentPoolKey.currency0, currency0)
+                    || !currencyEquals(currentPoolKey.currency1, currency1)
+            ) revert InvalidPoolKey();
+            PoolId poolId = currentPoolKey.toId();
+            balanceDeltaOfPool[poolId] = delta;
+        }
 
-        _accountDeltaForApp(address(key.poolManager), key.currency0, delta.amount0());
-        _accountDeltaForApp(address(key.poolManager), key.currency1, delta.amount1());
+        _accountDeltaForApp(currency0, delta.amount0());
+        _accountDeltaForApp(currency1, delta.amount1());
     }
 
-    function _accountDeltaForApp(address poolManager, Currency currency, int128 delta) internal {
```

Comment:
i prefer if this can be reverted, following Vault.sol. so developers might not spend a lot more effort figuring why the implementation is different 

```
// from Vault.sol
    function _accountDeltaForApp(address app, Currency currency, int128 delta) internal {
        if (delta == 0) return;

        if (delta >= 0) {
            /// @dev arithmetic underflow make sure trader can't withdraw too much from app
            reservesOfApp[app][currency] -= uint128(delta);
        } else {
            /// @dev arithmetic overflow make sure trader won't deposit too much into app
            reservesOfApp[app][currency] += uint128(-delta);
        }
    }
```